activejob/lib/active_job/exceptions.rb

```diff
@@ -0,0 +1,76 @@
+module ActiveJob
+  # Provides behavior for retrying and discarding jobs on exceptions.
+  module Exceptions
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      # Catch the exception and reschedule job for re-execution after so many seconds, for a specific number of attempts.
+      # If the exception keeps getting raised beyond the specified number of attempts, the exception is allowed to
+      # bubble up to the underlying queuing system, which may have its own retry mechanism or place it in a
+      # holding queue for inspection.
+      #
+      # ==== Options
+      # * <tt>:wait</tt> - Re-enqueues the job with the specified delay in seconds
+      # * <tt>:attempts</tt> - Re-enqueues the job the specified number of times
+      #
+      # ==== Examples
+      #
+      #  class RemoteServiceJob < ActiveJob::Base
+      #    retry_on Net::OpenTimeout, wait: 30.seconds, attempts: 10
+      #
+      #    def perform(*args)
+      #      # Might raise Net::OpenTimeout when the remote service is down
+      #    end
+      #  end
+      def retry_on(exception, wait: 3.seconds, attempts: 5)
+        rescue_from exception do |error|
+          logger.error "Retrying #{self.class} in #{wait} seconds, due to a #{exception}. The original exception was #{error.cause.inspect}."
+          retry_job wait: wait if executions < attempts
```

Comment:
Why pass on only `:wait` here? Since `retry_job` has an example of `rescue_from` that re-queues on another `queue`, maybe it's not that much of stretch that people might want to do that? I'd say the same goes for `priority`, though I can't see `wait_until` make sense as the time would be fixed.